Renames files to specify that they are separate for tests and experiments

jeppe-style opened the pull request:
Closes https://github.com/benchflow/benchflow/issues/124

jeppe-style committed:
Renames some files to specify that they are separate for tests and experiments.

jeppe-style committed:
Merge branch 'devel' of https://github.com/benchflow/benchflow into feature-dsl-refactor-filenames

* 'devel' of https://github.com/benchflow/benchflow:
  Renames a section
  Adds draft structure of developer guide
  Adds draft structure of getting started and user guide

jeppe-style committed:
Documents DSL library file naming conventions.

VincenzoFerme commented on docs/developer-guide/architecture-and-design.md:
```diff
@@ -19,4 +19,10 @@ group: developer-guide
 
-<!--  -->
\ No newline at end of file
+## Serializing/Deserializing YAML
```
This should be `###`, to be correctly placed under the `BenchFlow DSL` section.

VincenzoFerme commented on docs/developer-guide/architecture-and-design.md:
```diff
@@ -19,4 +19,10 @@ group: developer-guide
+To serialize/deserialize YAML the DSL library uses [MoultingYAML](https://github.com/jcazevedo/moultingyaml).
+
+### Naming of `*YamlProtocol` files and case classes
+Most of the `*YamlProtocol` files can be shared for both Tests and Experiments, as well as their case class counterparts. In the cases where we need to separate the protocols we different the files by prepending `BenchFlowTest` and `BenchFlowExperiment` for tests and experiments respectively, e.g. `BenchFlowExperimentConfigurationYamlProtocol` and `BenchFlowExperimentConfiguration`.
```
Make it also explicit that nothing is prepended when there is not need to separate files, because they are shared exactly the same between Test and Experiment.

jeppe-style committed:
Updates architecture-and-design.md

with changes as suggested in https://github.com/benchflow/benchflow/pull/126#pullrequestreview-31499117
